requestmanager/executor/executor.go

```diff
@@ -18,183 +19,172 @@ import (
 type AsyncLoadFn func(peer.ID, graphsync.RequestID, ipld.Link, ipld.LinkContext) <-chan types.AsyncLoadResult
 
-// ExecutionEnv are request parameters that last between requests
-type ExecutionEnv struct {
-	Ctx              context.Context
-	SendRequest      func(peer.ID, gsmsg.GraphSyncRequest)
-	RunBlockHooks    func(p peer.ID, response graphsync.ResponseData, blk graphsync.BlockData) error
-	TerminateRequest func(graphsync.RequestID)
-	WaitForMessages  func(ctx context.Context, resumeMessages chan graphsync.ExtensionData) ([]graphsync.ExtensionData, error)
-	Loader           AsyncLoadFn
-	LinkSystem       ipld.LinkSystem
+type Executor struct {
+	manager    Manager
+	blockHooks BlockHooks
+	loader     AsyncLoadFn
 }
 
-// RequestExecution are parameters for a single request execution
-type RequestExecution struct {
-	Ctx                  context.Context
-	P                    peer.ID
-	TerminalError        chan error
-	Request              gsmsg.GraphSyncRequest
-	LastResponse         *atomic.Value
-	DoNotSendCids        *cid.Set
-	NodePrototypeChooser traversal.LinkTargetNodePrototypeChooser
-	ResumeMessages       chan []graphsync.ExtensionData
-	PauseMessages        chan struct{}
+func NewExecutor(
+	manager Manager,
+	blockHooks BlockHooks,
+	loader AsyncLoadFn) *Executor {
+	return &Executor{
+		manager:    manager,
+		blockHooks: blockHooks,
+		loader:     loader,
+	}
 }
 
-// Start begins execution of a request in a go routine
-func (ee ExecutionEnv) Start(re RequestExecution) (chan graphsync.ResponseProgress, chan error) {
-	executor := &requestExecutor{
-		inProgressChan:   make(chan graphsync.ResponseProgress),
-		inProgressErr:    make(chan error),
-		ctx:              re.Ctx,
-		p:                re.P,
-		terminalError:    re.TerminalError,
-		request:          re.Request,
-		lastResponse:     re.LastResponse,
-		doNotSendCids:    re.DoNotSendCids,
-		nodeStyleChooser: re.NodePrototypeChooser,
-		resumeMessages:   re.ResumeMessages,
-		pauseMessages:    re.PauseMessages,
-		env:              ee,
+func (e *Executor) ExecuteTask(ctx context.Context, pid peer.ID, task *peertask.Task) bool {
+	requestTaskChan := make(chan RequestTask)
+	var requestTask RequestTask
+	e.manager.GetRequestTask(pid, task, requestTaskChan)
+	select {
+	case requestTask = <-requestTaskChan:
+	case <-ctx.Done():
+		return true
+	}
+	if requestTask.Empty {
+		log.Info("Empty task on peer request stack")
+		return false
 	}
-	executor.sendRequest(executor.request)
-	go executor.run()
-	return executor.inProgressChan, executor.inProgressErr
+	log.Debugw("beginning request execution", "id", requestTask.Request.ID(), "peer", pid.String(), "root_cid", requestTask.Request.Root().String())
+	err := e.traverse(requestTask)
+	if err != nil {
+		if !isContextErr(err) {
+			e.manager.SendRequest(requestTask.P, gsmsg.CancelRequest(requestTask.Request.ID()))
+		}
+		if !isContextErr(err) && !isPausedErr(err) {
```

Comment:
could skip `isContextErr` check here by nesting in the above block, but also is this logic of these two blocks and order correct here? The previous form in `run` doesn't check `isPausedErr` and it seems strange to skip passing in the error if that's the case.

Reply:
The previous form would not end on IsPausedError -- it handled the pause/resume internally and the request all the way to the end.